backend/de.metas.util.web/src/main/java/de/metas/util/web/audit/ApiAuditService.java

```diff
@@ -0,0 +1,345 @@
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.function.Supplier;
+
+@Service
+public class ApiAuditService
+{
+	public static final String API_FILTER_REQUEST_ID_HEADER = "X-ApiFilter-Request-ID";
+
+	private final ApiAuditConfigRepository apiAuditConfigRepository;
+	private final ApiRequestAuditRepository apiRequestAuditRepository;
+	private final ApiResponseAuditRepository apiResponseAuditRepository;
+	private final ApiAuditRequestLogDAO apiAuditRequestLogDAO;
+
+	private final WebClient webClient;
+	private final ConcurrentHashMap<UserId, HttpCallScheduler> callerId2Scheduler;
+	private final ObjectMapper objectMapper;
+
+	public ApiAuditService(
+			@NonNull final ApiAuditConfigRepository apiAuditConfigRepository,
+			@NonNull final ApiRequestAuditRepository apiRequestAuditRepository,
+			@NonNull final ApiResponseAuditRepository apiResponseAuditRepository,
+			@NonNull final ApiAuditRequestLogDAO apiAuditRequestLogDAO)
+	{
+		this.apiAuditConfigRepository = apiAuditConfigRepository;
+		this.apiRequestAuditRepository = apiRequestAuditRepository;
+		this.apiResponseAuditRepository = apiResponseAuditRepository;
+		this.apiAuditRequestLogDAO = apiAuditRequestLogDAO;
+		this.webClient = WebClient.create();
+		this.callerId2Scheduler = new ConcurrentHashMap<>();
+		this.objectMapper = JsonObjectMapperHolder.newJsonObjectMapper();
+	}
+
+	public boolean wasAlreadyFiltered(@NonNull final HttpServletRequest httpServletRequest)
+	{
+		return httpServletRequest.getHeader(API_FILTER_REQUEST_ID_HEADER) != null;
+	}
+
+	@NonNull
+	public Optional<ApiAuditConfig> getMatchingAuditConfig(@NonNull final ServletRequest request)
+	{
+		final CustomHttpRequestWrapper requestWrapper = new CustomHttpRequestWrapper((HttpServletRequest)request, objectMapper);
+
+		if (Check.isBlank(requestWrapper.getFullPath()) || Check.isBlank(requestWrapper.getHttpMethodString()))
+		{
+			return Optional.empty();
+		}
+
+		final OrgId orgId = Env.getOrgId();
+		final ImmutableList<ApiAuditConfig> apiAuditConfigs = apiAuditConfigRepository.getAllConfigsByOrgId(orgId);
+
+		return apiAuditConfigs.stream()
+				.filter(config -> config.matchesRequest(requestWrapper.getFullPath(), requestWrapper.getHttpMethodString()))
+				.min(Comparator.comparingInt(ApiAuditConfig::getSeqNo));
+	}
+
+	public void processHttpCall(
+			@NonNull final HttpServletRequest request,
+			@NonNull final HttpServletResponse response,
+			@NonNull final ApiAuditConfig apiAuditConfig)
+	{
+		final OrgId orgId = apiAuditConfig.getOrgId();
+
+		final CustomHttpRequestWrapper httpRequest = new CustomHttpRequestWrapper(request, objectMapper);
+
+		final ApiRequestAudit requestAudit = logRequest(httpRequest, apiAuditConfig.getApiAuditConfigId(), orgId);
+
+		final ApiAuditLoggable apiAuditLoggable = createLogger(requestAudit.getIdNotNull());
+
+		try (final IAutoCloseable loggableRestorer = Loggables.temporarySetLoggable(apiAuditLoggable))
```

Comment:
i didn't yet dig into CompletableFuture, but are we sure that `apiAuditLoggable` will be in the threat in which the request will be processed? on my first glimpse i think the executor in HttpCallScheduler won't have the loggable..